ACD-29 Enhanced security check to account for inherited/combined access control context

paouelle opened the pull request:
### Description of the Change
Enhanced security check to account for inherited/combined access control context when computing.
Made sure to escape quoted strings when generating a permission string for the policy file.
Enhanced the backdoor to replace path separators of file permissions with the special ${/} entry if the corresponding system property is defined.
Added new --reconnect and --fail options to support being used with itests.

### Benefits
Ensures that the AC debugger will be able to account for scenarios where security failures are generated not because of bundles in the stack missing permissions but because inherited bundles don't have the permissions required.

### Verification Process
0. Modify DDF main pom to reference the same ac debugger version as built from this PR
1. Build DDF
2. Install DDF
3. Replace the `<<ALL FILES>>` for the ssh bundles:
```
grant codeBase "file:/org.apache.karaf.shell.ssh/org.apache.sshd.core" {
    permission java.io.FilePermission "<<ALL FILES>>", "read,write";
}
```
with
```
grant codeBase "file:/org.apache.karaf.shell.ssh/org.apache.sshd.core" {
    permission java.io.FilePermission "${ddf.home.perm}etc${/}keystores${/}-", "read";
    permission java.io.FilePermission "${ddf.home.perm}etc${/}ws-security${/}server${/}-", "read";
    permission java.io.FilePermission "${ddf.home.perm}etc${/}host.key", "read,write";
    permission java.io.FilePermission "${ddf.home.perm}etc${/}branding-ssh.properties", "read";
    permission java.io.FilePermission "${ddf.home.perm}etc${/}org.apache.karaf.shell.cfg", "read";
}
```
4. Connect to DDF using an SSH session: `ssh -p 8101 -o PasswordAuthentication=yes -o PubkeyAuthentication=no -vv  admin@localhost`
5. Start the AC debugger with --debug --continuous --wait
6. Attempt to ingest a file not located under DDF.HOME or the user's home directory via the SSH console

### Applicable Issues
Fixes: #29

paouelle committed:
ACD-29 Enhanced security check to account for inherited/combined access control context when computing.

Made sure to escape quoted strings when generating a permission string for the policy file.
Enhanced the backdoor to replace path separators of file permissions with the special ${/} entry if defined.

paouelle commented on common/src/main/java/org/codice/acdebugger/common/PermissionUtil.java:
```diff
@@ -0,0 +1,103 @@
+   * @param str the string to be quoted
+   * @return the corresponding quoted string
+   */
+  private static String escape(String str) {
```
ℹ️ We are implementing the counterpart to the StreamTokenizer class used by the policy parser to escape special characters. I needed to avoid pulling in more dependencies such as Apache's commons-text which has an implementation that might work as this code is also used by the backdoor bundle that is loaded as part of the DDF kernel.

paouelle commented:
build now

coyotesqrl commented on debugger/src/main/java/org/codice/acdebugger/breakpoints/SecurityCheckInformation.java:
```diff
@@ -67,8 +68,12 @@
+
+  /** the current loop index through the domain context in the code where the breakpoint occurred */
+  @SuppressWarnings("squid:S00116" /* name is clearer that way */)
+  private final int local_i;
```
✏️ Could try a name like `loopIndex` or `localLoopIdx` instead. Those are probably as clear.

paouelle replied:
I wanted to use i as in the AccessController's code, the variable is named `i`

coyotesqrl replied:
That's perfectly okay by me.

coyotesqrl commented on debugger/src/main/java/org/codice/acdebugger/breakpoints/SecurityCheckInformation.java:
```diff
@@ -192,10 +206,19 @@ private SecurityCheckInformation(
       // since bundle-0 always has all permissions, we cannot received null as the current bundle
       // where the failure occurred
-      dumpTroubleshootingInfo(local_i);
+      dumpTroubleshootingInfo(
```
❓ WHY ARE YOU YELLING? 😄

paouelle replied:
😄 Not intended to be *yelling*. The intent is to make this information stand out in the output. Do you think I should not be doing upper case?

coyotesqrl replied:
I'm on the fence about it. Walls of uppercase text are generally harder to read but they can provide important emphasis. I really don't care either way.

paouelle committed:
ACD-29 Fixed issue with combined domains which would typically go away from a new access control context that results from adding a doPrivileged() block.

Fix sorting issue with solutions.
Fix issue where during analysis, it wasn't ignoring solutions which were granting permissions to too many bundles.

paouelle committed:
ACD-29 Added new --reconnect and --fail options.

Fix sleep between attempt to connect which was no longer waiting.

paouelle committed:
ACD-29 Added support for retrieving bundle versions in backdoor

figliold commented on debugger/src/main/java/org/codice/acdebugger/ACDebugger.java:
```diff
@@ -71,6 +69,13 @@
 
+  @Option(
+    names = {"-f", "--fail"},
+    description =
```
✏️ Seems like there are some extra words in the description.

paouelle committed:
ACD-29 fix option description
